chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java

```diff
@@ -0,0 +1,307 @@
+    private boolean registered = false;
+    private StatusExchangeThread statusExchangeThread;
+    
+    public MasterToJaasCoordinator(String envId, String jaasEndpoint, int statusReportIntervalSeconds,
+                                   Set<String> availableConfigs) {
+        this.envId = envId;
+        try {
+            this.jaasEndpoint = new URI(jaasEndpoint);
+        } catch (URISyntaxException e) {
+            throw new IllegalStateException("Incorrect JaaS endpoint", e);
+        }
+    
+        if (availableConfigs == null) {
+            availableConfigs = Collections.emptySet();
+        }
+        this.availableConfigs = availableConfigs;
+        
+        if (statusReportIntervalSeconds <= 0) {
+            LOGGER.warn(
+                    "Provided status reporting interval seconds is less or equals to zero - {}.\n"
+                    + "Using the default value - 15 seconds",
+                    statusReportIntervalSeconds
+            );
+            statusReportIntervalSeconds = 15;
+        }
+        this.statusReportIntervalSeconds = statusReportIntervalSeconds;
+    
+    }
+    
+    public void register() {
+        final String sessionCookie = doRegister();
+        statusExchangeThread = new StatusExchangeThread(sessionCookie);
+        statusExchangeThread.start();
+        registered = true;
+    }
+    
+    public String awaitConfigToExecute() throws TerminateException {
+        if (!registered) {
+            throw new IllegalStateException("must be registered");
+        }
+        try {
+            statusExchangeThread.setPendingRequestEntity();
+            String configName;
+            do {
+                configName = statusExchangeThread.nextConfigToExecute.poll(1, TimeUnit.MINUTES);
+                if (!standBy) {
+                    throw new TerminateException("Master execution can't be proceeded");
+                }
+            } while (configName == null);
+            statusExchangeThread.setRunningRequestEntity(configName);
+            return configName;
+        } catch (InterruptedException e) {
+            statusExchangeThread.interrupt();
+            throw new TerminateException("Master execution can't be proceeded");
+        }
+    }
+    
+    private String doRegister() {
+        URI envsUri = UriComponentsBuilder.newInstance().uri(jaasEndpoint).path("/envs").build().toUri();
+    
+        TestEnvironmentEntity testEnvironmentEntity = buildTestEnvironmentEntityWith(TestEnvironmentStatus.PENDING);
+        ResponseEntity<String> responseEntity = null;
+        boolean posted = false;
+        do {
+            try {
+                responseEntity = restTemplate.postForEntity(envsUri, testEnvironmentEntity, String.class);
+                posted = true;
+            } catch (HttpClientErrorException e) {
+                if (!isEnvIdUnacceptable(e)) {
+                    throw e;
+                }
+                envId = UUID.randomUUID().toString();
```

Comment:
One of the reasons we got HttpClientErrorException  here - ID doesn't match to the pattern.